// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: mesh/v1alpha1/externalservice.proto

package v1alpha1

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"strings"
	"time"
	"unicode/utf8"

	"github.com/golang/protobuf/ptypes"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = ptypes.DynamicAny{}
)

// define the regex for a UUID once up-front
var _externalservice_uuidPattern = regexp.MustCompile("^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$")

// Validate checks the field values on ExternalService with the rules defined
// in the proto definition for this message. If any rules are violated, an
// error is returned.
func (m *ExternalService) Validate() error {
	if m == nil {
		return nil
	}

	if v, ok := interface{}(m.GetNetworking()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return ExternalServiceValidationError{
				field:  "Networking",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if v, ok := interface{}(m.GetMetrics()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return ExternalServiceValidationError{
				field:  "Metrics",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	return nil
}

// ExternalServiceValidationError is the validation error returned by
// ExternalService.Validate if the designated constraints aren't met.
type ExternalServiceValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ExternalServiceValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ExternalServiceValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ExternalServiceValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ExternalServiceValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ExternalServiceValidationError) ErrorName() string { return "ExternalServiceValidationError" }

// Error satisfies the builtin error interface
func (e ExternalServiceValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sExternalService.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ExternalServiceValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ExternalServiceValidationError{}

// Validate checks the field values on ExternalService_Networking with the
// rules defined in the proto definition for this message. If any rules are
// violated, an error is returned.
func (m *ExternalService_Networking) Validate() error {
	if m == nil {
		return nil
	}

	// no validation rules for Address

	for idx, item := range m.GetInbound() {
		_, _ = idx, item

		if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ExternalService_NetworkingValidationError{
					field:  fmt.Sprintf("Inbound[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	return nil
}

// ExternalService_NetworkingValidationError is the validation error returned
// by ExternalService_Networking.Validate if the designated constraints aren't met.
type ExternalService_NetworkingValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ExternalService_NetworkingValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ExternalService_NetworkingValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ExternalService_NetworkingValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ExternalService_NetworkingValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ExternalService_NetworkingValidationError) ErrorName() string {
	return "ExternalService_NetworkingValidationError"
}

// Error satisfies the builtin error interface
func (e ExternalService_NetworkingValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sExternalService_Networking.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ExternalService_NetworkingValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ExternalService_NetworkingValidationError{}

// Validate checks the field values on ExternalService_Networking_Inbound with
// the rules defined in the proto definition for this message. If any rules
// are violated, an error is returned.
func (m *ExternalService_Networking_Inbound) Validate() error {
	if m == nil {
		return nil
	}

	// no validation rules for Address

	// no validation rules for Port

	// no validation rules for ServiceAddress

	// no validation rules for ServicePort

	if len(m.GetTags()) < 1 {
		return ExternalService_Networking_InboundValidationError{
			field:  "Tags",
			reason: "value must contain at least 1 pair(s)",
		}
	}

	return nil
}

// ExternalService_Networking_InboundValidationError is the validation error
// returned by ExternalService_Networking_Inbound.Validate if the designated
// constraints aren't met.
type ExternalService_Networking_InboundValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ExternalService_Networking_InboundValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ExternalService_Networking_InboundValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ExternalService_Networking_InboundValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ExternalService_Networking_InboundValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ExternalService_Networking_InboundValidationError) ErrorName() string {
	return "ExternalService_Networking_InboundValidationError"
}

// Error satisfies the builtin error interface
func (e ExternalService_Networking_InboundValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sExternalService_Networking_Inbound.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ExternalService_Networking_InboundValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ExternalService_Networking_InboundValidationError{}
